source/common/ext_authz/ext_authz_impl.cc

```diff
@@ -0,0 +1,218 @@
+#include "common/ext_authz/ext_authz_impl.h"
+
+#include <chrono>
+#include <cstdint>
+#include <string>
+#include <vector>
+
+#include "envoy/access_log/access_log.h"
+#include "envoy/ssl/connection.h"
+
+#include "common/common/assert.h"
+#include "common/grpc/async_client_impl.h"
+#include "common/http/headers.h"
+#include "common/protobuf/protobuf.h"
+
+#include "fmt/format.h"
+
+namespace Envoy {
+namespace ExtAuthz {
+
+GrpcClientImpl::GrpcClientImpl(Grpc::AsyncClientPtr&& async_client,
+                               const Optional<std::chrono::milliseconds>& timeout)
+    : service_method_(*Protobuf::DescriptorPool::generated_pool()->FindMethodByName(
+          "envoy.service.auth.v2.Authorization.Check")),
+      async_client_(std::move(async_client)), timeout_(timeout) {}
+
+GrpcClientImpl::~GrpcClientImpl() { ASSERT(!callbacks_); }
+
+void GrpcClientImpl::cancel() {
+  ASSERT(callbacks_ != nullptr);
+  request_->cancel();
+  callbacks_ = nullptr;
+}
+
+void GrpcClientImpl::check(RequestCallbacks& callbacks,
+                           const envoy::service::auth::v2::CheckRequest& request,
+                           Tracing::Span& parent_span) {
+  ASSERT(callbacks_ == nullptr);
+  callbacks_ = &callbacks;
+
+  request_ = async_client_->send(service_method_, request, *this, parent_span, timeout_);
+}
+
+void GrpcClientImpl::onSuccess(std::unique_ptr<envoy::service::auth::v2::CheckResponse>&& response,
+                               Tracing::Span& span) {
+  CheckStatus status = CheckStatus::OK;
+  ASSERT(response->status().code() != Grpc::Status::GrpcStatus::Unknown);
+  if (response->status().code() != Grpc::Status::GrpcStatus::Ok) {
+    status = CheckStatus::Denied;
+    span.setTag(Constants::get().TraceStatus, Constants::get().TraceUnauthz);
+  } else {
+    span.setTag(Constants::get().TraceStatus, Constants::get().TraceOk);
+  }
+
+  callbacks_->onComplete(status);
+  callbacks_ = nullptr;
+}
+
+void GrpcClientImpl::onFailure(Grpc::Status::GrpcStatus status, const std::string&,
+                               Tracing::Span&) {
+  ASSERT(status != Grpc::Status::GrpcStatus::Ok);
+  UNREFERENCED_PARAMETER(status);
+  callbacks_->onComplete(CheckStatus::Error);
+  callbacks_ = nullptr;
+}
+
+void ExtAuthzCheckRequestGenerator::addressToProtobufAddress(
```

Comment:
Should this be a feature of `Network::Address::Instance` or in https://github.com/envoyproxy/envoy/blob/master/source/common/network/utility.h? Seems generally useful.